### PR TITLE
Load add-on returned by the API

### PR DESCRIPTION
**Pull request by willdurand**
Fixes https://github.com/mozilla/addons-frontend/issues/9731

### CODE REVIEW

File: src/amo/pages/Addon/index.js [hunk 1]
```diff
@@ -109,13 +101,15 @@ export class AddonBase extends React.Component {
     // This makes sure we do not try to dispatch any new actions in the case
     // of an error.
     if (!errorHandler.hasError()) {
-      const { slug } = params;
```
Comment (bobsilverberg) [inline src/amo/pages/Addon/index.js:112]: Why did you remove this? It just means you have to use `params.slug` twice below, so it seems to have introduced churn without any real benefit.

File: src/core/reducers/addons.js [hunk 2]
```diff
@@ -41,15 +41,17 @@ export type AddonsState = {|
   infoBySlug: {
     [slug: string]: {| info: AddonInfoType, loading: boolean |},
   },
-  loadingBySlug: { [addonSlug: string]: boolean },
+  byIdInURL: { [addonSlug: string]: AddonID },
```
Comment (bobsilverberg) [inline src/core/reducers/addons.js:44]: Should this be renamed to `[addonId: string]`?

File: src/core/reducers/addons.js [hunk 3]
```diff
@@ -261,7 +255,7 @@ export const isAddonLoading = (state: AppState, slug: string): boolean => {
     return false;
   }
 
-  return Boolean(state.addons.loadingBySlug[slug.toLowerCase()]);
+  return Boolean(state.addons.loadingByIdInURL[slug]);
```
Comment (bobsilverberg) [inline src/core/reducers/addons.js:258]: Do you think this instance of `slug` should be renamed to `addonId`?

File: src/core/reducers/addons.js [hunk 4]
```diff
@@ -380,9 +380,9 @@ export default function addonsReducer(
             ...state.bySlug,
             [addon.slug.toLowerCase()]: undefined,
           },
-          loadingBySlug: {
-            ...state.loadingBySlug,
-            [addon.slug.toLowerCase()]: undefined,
+          loadingByIdInURL: {
+            ...state.loadingByIdInURL,
+            [addon.slug]: undefined,
```
Comment (bobsilverberg) [inline src/core/reducers/addons.js:385]: Does this conflict with the other logic in this patch? I guess we don't have a `slug` to pass into this action like we do with `loadAddon`, and I guess it doesn't really matter.

File: tests/unit/amo/pages/TestAddon.js [hunk 5]
```diff
@@ -1612,10 +1603,56 @@ describe(__filename, () => {
...
+      fakeDispatch,
+      sendServerRedirect({
+        status: 301,
+        url: `/en-US/${clientApp}${getAddonURL(addon.slug)}`,
+      }),
+    );
+    sinon.assert.callCount(fakeDispatch, 1);
+  });
+
+  it(`dispatches a server redirect when slug is the add-on's GUID using a different case`, () => {
```
Comment (bobsilverberg) [inline tests/unit/amo/pages/TestAddon.js:1644]: This test case seems unnecessary. We've already tested with a GUID, and a GUID with a different case shouldn't be treated any differently, afaict.

File: tests/unit/core/reducers/test_addons.js [hunk 6]
```diff
@@ -277,28 +296,22 @@ describe(__filename, () => {
...
 
   describe('loadAddon', () => {
-    it('requires addons', () => {
+    it('requires an addon', () => {
       expect(() => {
         loadAddon();
       }).toThrow('addon is required');
     });
+
+    it('requires a slug', () => {
```
Comment (bobsilverberg) [inline tests/unit/core/reducers/test_addons.js:310]: We generally don't add tests for `invariant`s. I know there was one already for "requires an addon", but maybe we'd be better removing that one, than adding another?

File: tests/unit/core/reducers/test_addons.js [hunk 7]
```diff
@@ -310,78 +323,14 @@ describe(__filename, () => {
...
-      store.dispatch(loadAddon({ addon: fakeAddon }));
+      store.dispatch(loadAddon({ addon: fakeAddon, slug: fakeAddon.slug }));
 
       expect(getAddonByID(store.getState().addons, fakeAddon.id)).toEqual(
         createInternalAddon(fakeAddon),
       );
     });
   });
 
-  describe('getAddonBySlug', () => {
```
Comment (bobsilverberg) [inline tests/unit/core/reducers/test_addons.js:321]: Do we need any tests for the new `getAddonByIdInURL` function?

**Review by bobsilverberg — APPROVED** [review]
Looks good, thanks for taking this on @willdurand. I added a number of comments, but they are all minor.

r+wc

File: src/amo/pages/Addon/index.js (hunk 1 above)
Comment (willdurand) [reply src/amo/pages/Addon/index.js:112]: I find it more readable because `params.slug` might not be a `slug`. Having `params.` makes it easier to me to understand that it's the name of the parameter in the URL.

File: src/core/reducers/addons.js (hunk 2 above)
Comment (willdurand) [reply src/core/reducers/addons.js:44]: probably yes

File: src/core/reducers/addons.js (hunk 3 above)
Comment (willdurand) [reply src/core/reducers/addons.js:258]: yes, thanks

File: src/core/reducers/addons.js (hunk 4 above)
Comment (willdurand) [reply src/core/reducers/addons.js:385]: yeah, I think so. No test breaks so I guess it doesn't really matter as you said.

File: tests/unit/core/reducers/test_addons.js (hunk 6 above)
Comment (willdurand) [reply tests/unit/core/reducers/test_addons.js:310]: yep I am fine with that

File: tests/unit/core/reducers/test_addons.js (hunk 7 above)
Comment (willdurand) [reply tests/unit/core/reducers/test_addons.js:321]: ha maybe yes
